### PR TITLE
DDC-2842: Leave out discriminator part of Doctrine' generated SQL on root entity

### PR DESCRIPTION
**Pull request by 11ya**
Fix for [#2842](https://github.com/doctrine/doctrine2/issues/3596)

### CODE REVIEW

**Review by Ocramius — CHANGES_REQUESTED** [review]
The reason why the `t.discriminator_column IN()` condition is added is that we do not assume totality of record ownership on that table.

Records may have an unknown discriminator or belong to other inheritance trees.

While this change is possibly an improvement, it also represents a BC break, as hydration would just choke when encountering an unknown discriminator value.

**Comment by 11ya** [comment]
@Ocramius Ok, if I understand you correctly, check on root entity class name should solve the problem of "belong to other inheritance trees". As for "unknown discriminator", I see only one decision - add checking discriminator value constraint to the table definition and separate option in discriminator map for that.

**Comment by Ocramius** [comment]
> I see only one decision - add checking discriminator value constraint to the table definition and separate option in discriminator map for that.

No way: this is a lot of added complexity for something that DB indexes solved already

File: lib/Doctrine/ORM/Mapping/DiscriminatorColumn.php [hunk 1]
```diff
@@ -51,4 +51,9 @@
      * @var string
      */
     public $columnDefinition;
+
+    /**
+     * @var bool
+     */
+    public $strict;
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Mapping/DiscriminatorColumn.php]: As I mentioned before, an addition here is not to be considered unless an actual problem is being solved by it. So far, it seems like this only leads to removal of an `IN()` condition on discrimitor values when set to `false`

File: lib/Doctrine/ORM/Tools/SchemaTool.php [hunk 2]
```diff
@@ -394,7 +394,15 @@ private function addDiscriminatorColumnDefinition($class, Table $table)
...
 
-        $table->addColumn($discrColumn['name'], $discrColumn['type'], $options);
+        $column = $table->addColumn($discrColumn['name'], $discrColumn['type'], $options);
+
+        if (isset($discrColumn['strict']) && $discrColumn['strict']) {
+            $values = [];
+            foreach ($class->discriminatorMap as $value => $className) {
+                $values[] = $this->platform->quoteStringLiteral($value);
+            }
+            $column->setCustomSchemaOption('check', 'CHECK (' . $discrColumn['name'] . ' IN (' . implode(',', $values) . '))');
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/SchemaTool.php]: Don't think we'll manage to make this work cross-platform

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
Taking a decision not to pursue this further: there is some improvement in the generated SQL, yes, but most decent SQL query optimisers can deal with these details anyway.

Therefore, closing as `won't fix` here, sorry @11ya!

**Pull request closed without merging**